### PR TITLE
Add a new module to manage Virtual Port Channel Interfaces in Fabric Resource Policies Template. (DCNE-53)

### LINKED ISSUE #467: Enhancement: New module for Virtual Port Channel Interface in Fabric Resource Policies Template (DCNE-53)
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Create new module for Virtual Port Channel Interface in Fabric Resource Policies Template ( created via ndo_template module )

### New or Affected Module(s):

* mso_XXXXX

### MSO version and MSO Platform

* ND 3.1.1 / NDO 4.3.1

### APIC version and APIC Platform for Site Level Resources

* V x.x.x and on-prem/cloud-aws/cloud-azure/all.

### Collection versions

* cisco.mso x.x.x

### References

* #0000

### PR DESCRIPTION
**Pull request by gmicol**
resolves #467 

New Modules Added:
- ndo_virtual_port_channel_interface

### CODE REVIEW

File: plugins/modules/ndo_virtual_port_channel_interface.py [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+        aliases: [ policy, interface_policy ]
+    interface_descriptions:
+        description:
+        - The list of descriptions for each interface.
+        type: list
+        elements: dict
+        suboptions:
+            node:
+                description:
+                - The node ID
```
Comment (akinross) [inline plugins/modules/ndo_virtual_port_channel_interface.py:99]:
```suggestion
                - The node ID.
```

File: plugins/modules/ndo_virtual_port_channel_interface.py [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+        type: str
+    node_1:
+        description:
+        - The first node ID.
+        type: str
+    node_2:
+        description:
+        - The second node ID.
+        type: str
+    interfaces_1:
```
[the remark this answers is not in the document]
Comment (gmicol) [reply plugins/modules/ndo_virtual_port_channel_interface.py:57]: I like this idea, I will ad an alias as well

File: plugins/modules/ndo_virtual_port_channel_interface.py [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+        type: str
+        required: true
+    virtual_port_channel_interface:
+        description:
+        - The name of the Virtual Port Channel Interface.
+        type: str
+        aliases: [ name, virtual_port_channel, vpc ]
+    virtual_port_channel_interface_uuid:
+        description:
+        - The uuid of the Virtual Port Channel Interface.
```
[the remark this answers is not in the document]
Comment (gmicol) [reply plugins/modules/ndo_virtual_port_channel_interface.py:41]: will change it, thanks

File: plugins/modules/ndo_virtual_port_channel_interface.py [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+        - This parameter can be used instead of O(interface_policy_group_uuid).
+        type: dict
+        suboptions:
+            name:
+                description:
+                - The name of the Port Channel Interface Setting Policy.
+                type: str
+            template:
+                description:
+                - The name of the template in which is referred the Port Channel Interface Policy Group.
```
Comment (akinross) [inline plugins/modules/ndo_virtual_port_channel_interface.py:88]: this sentence is a bit odd for me to read, could you rephrase?

File: plugins/modules/ndo_virtual_port_channel_interface.py [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+                type: str
+        aliases: [ policy, interface_policy ]
+    interface_descriptions:
+        description:
+        - The list of descriptions for each interface.
```
Comment (akinross) [inline plugins/modules/ndo_virtual_port_channel_interface.py:93]:
```suggestion
        - The list of interface descriptions.
```

File: plugins/modules/ndo_virtual_port_channel_interface.py [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+        - Use C(query) for listing an object or multiple objects.
+        - Use C(present) for creating or updating.
+        type: str
+        choices: [ absent, query, present ]
+        default: query
+extends_documentation_fragment: cisco.mso.modules
+"""
+
+EXAMPLES = r"""
+- name: Create a new Virtual Port Channel Interface
```
Comment (akinross) [inline plugins/modules/ndo_virtual_port_channel_interface.py:121]: examples for uuid update/query/delete are missing

File: plugins/modules/ndo_virtual_port_channel_interface.py [hunk 7]
```diff
@@ -0,0 +1,383 @@
...
+            fabric_policy_template = MSOTemplate(mso, "fabric_policy", interface_policy_group.get("template"))
+            fabric_policy_template.validate_template("fabricPolicy")
+            interface_policy_group_uuid = fabric_policy_template.get_interface_policy_group_uuid(interface_policy_group.get("name"))
+
+        if match:
+            if virtual_port_channel_interface and match.details.get("name") != virtual_port_channel_interface:
+                ops.append(dict(op="replace", path="{0}/{1}/name".format(path, match.index), value=virtual_port_channel_interface))
+                match.details["name"] = virtual_port_channel_interface
+
+            if description and match.details.get("description") != description:
```
Comment (akinross) [inline plugins/modules/ndo_virtual_port_channel_interface.py:290]: how do we rest to None? should this be 
```suggestion
            if description is not None and match.details.get("description") != description:
```

File: plugins/modules/ndo_virtual_port_channel_interface.py [hunk 8]
```diff
@@ -0,0 +1,383 @@
...
+    argument_spec = mso_argument_spec()
+    argument_spec.update(
+        template=dict(type="str", required=True),
+        virtual_port_channel_interface=dict(type="str", aliases=["name", "virtual_port_channel", "vpc"]),
+        virtual_port_channel_interface_uuid=dict(type="str", aliases=["uuid", "virtual_port_channel_uuid", "vpc_uuid"]),
+        description=dict(type="str"),
+        node_1=dict(type="str"),
+        node_2=dict(type="str"),
+        interfaces_node_1=dict(type="list", elements="str", aliases=["interfaces_1", "members_1"]),
+        interfaces_node_2=dict(type="list", elements="str", aliases=["interfaces_2", "members_2"]),
```
Comment (akinross) [inline plugins/modules/ndo_virtual_port_channel_interface.py:208]: should these not be required? how would you create a new vpc without node/interface config? for the update scenario I can understand it can be not provided and that the update is then ignored. else we expect a removal operation is also possible?

File: plugins/modules/ndo_virtual_port_channel_interface.py (hunk 7 above)
Comment (akinross) [reply plugins/modules/ndo_virtual_port_channel_interface.py:290]: same would apply for any other string that could be reset to ""

File: plugins/modules/ndo_virtual_port_channel_interface.py [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+
+            if interfaces_node_1 and interfaces_node_1 != match.details.get("node1Details", {}).get("memberInterfaces"):
+                ops.append(dict(op="replace", path="{0}/{1}/node1Details/memberInterfaces".format(path, match.index), value=interfaces_node_1))
+                match.details["node1Details"]["memberInterfaces"] = interfaces_node_1
+
+            if interfaces_node_2 and interfaces_node_2 != match.details.get("node2Details", {}).get("memberInterfaces"):
+                ops.append(dict(op="replace", path="{0}/{1}/node2Details/memberInterfaces".format(path, match.index), value=interfaces_node_2))
+                match.details["node2Details"]["memberInterfaces"] = interfaces_node_2
+
+            if interface_descriptions:
```
Comment (akinross) [inline plugins/modules/ndo_virtual_port_channel_interface.py:314]: just confirmation can we set descriptions to node that are not part of the VPV or should we validate this? or does ndo error in this scenario?

**Review by akinross — CHANGES_REQUESTED** [review]

File: plugins/modules/ndo_virtual_port_channel_interface.py [hunk 10]
```diff
@@ -0,0 +1,383 @@
...
+        interfaces_node_2=dict(type="list", elements="str", aliases=["interfaces_2", "members_2"]),
+        interface_policy_group=dict(
+            type="dict",
+            options=dict(
+                name=dict(type="str"),
+                template=dict(type="str"),
+            ),
+            aliases=["policy", "interface_policy"],
+        ),
+        interface_policy_group_uuid=dict(type="str", aliases=["policy_uuid", "interface_policy_uuid"]),
```
Comment (akinross) [inline plugins/modules/ndo_virtual_port_channel_interface.py:217]: i think one of these is also required on create

File: plugins/modules/ndo_virtual_port_channel_interface.py (hunk 8 above)
Comment (gmicol) [reply plugins/modules/ndo_virtual_port_channel_interface.py:208]: The API's errors out if they are missing during creation

File: plugins/modules/ndo_virtual_port_channel_interface.py (hunk 9 above)
Comment (gmicol) [reply plugins/modules/ndo_virtual_port_channel_interface.py:314]: Same thing, The API's errors out if they are matching what are the current nodes/interfaces config

File: plugins/modules/ndo_virtual_port_channel_interface.py (hunk 7 above)
Comment (gmicol) [reply plugins/modules/ndo_virtual_port_channel_interface.py:290]: you're right, will change that

File: plugins/modules/ndo_virtual_port_channel_interface.py (hunk 6 above)
Comment (gmicol) [reply plugins/modules/ndo_virtual_port_channel_interface.py:121]: will add them

File: plugins/modules/ndo_virtual_port_channel_interface.py (hunk 4 above)
Comment (gmicol) [reply plugins/modules/ndo_virtual_port_channel_interface.py:88]: sure, will rephrase it.

File: plugins/modules/ndo_virtual_port_channel_interface.py (hunk 5 above)
Comment (gmicol) [reply plugins/modules/ndo_virtual_port_channel_interface.py:93]: will change it

File: plugins/modules/ndo_virtual_port_channel_interface.py [hunk 11]
```diff
@@ -0,0 +1,383 @@
...
+        elements: str
+        aliases: [ interfaces_2, members_2 ]
+    interface_policy_group_uuid:
+        description:
+        - The UUID of the Port Channel Interface Setting Policy.
+        type: str
+        aliases: [ policy_uuid, interface_policy_uuid ]
+    interface_policy_group:
+        description:
+        - The name of the Port Channel Interface Policy Group.
```
[the remark this answers is not in the document]
Comment (gmicol) [reply plugins/modules/ndo_virtual_port_channel_interface.py:78]: what I could do is add an alias for it as well

**Review by akinross — APPROVED** [review]
LGTM

**Review by shrsr — CHANGES_REQUESTED** [review]

**Review by akinross — DISMISSED** [review]
LGTM

File: tests/integration/targets/ndo_virtual_port_channel_interface/tasks/main.yml [hunk 12]
```diff
@@ -0,0 +1,401 @@
...
+        state: query
+      register: query_all
+    
+    - name: Assert virtual port channel interface query tasks
+      assert:
+        that:
+          - query_one is not changed
+          - query_one.current.name == "ansible_virtual_port_channel_interface_2"
+          - query_all is not changed
+          - query_all.current | length == 2
```
Comment (samiib) [inline tests/integration/targets/ndo_virtual_port_channel_interface/tasks/main.yml:363]: Should you also check that the 2 interfaces are the expected 2 returned? Maybe check the names of both are correct.

**Review by samiib — CHANGES_REQUESTED** [review]

File: tests/integration/targets/ndo_virtual_port_channel_interface/tasks/main.yml (hunk 12 above)
Comment (gmicol) [reply tests/integration/targets/ndo_virtual_port_channel_interface/tasks/main.yml:363]: why not, I will add a few assessment for checking if they are the correct ones

**Review by shrsr — DISMISSED** [review]
LGTM

File: plugins/modules/ndo_virtual_port_channel_interface.py [hunk 13]
```diff
@@ -0,0 +1,412 @@
...
+    - If both parameter are used, O(interface_policy_group) will be ignored.
+    type: dict
+    suboptions:
+      name:
+        description:
+        - The name of the Port Channel Interface Policy Group.
+        type: str
+      template:
+        description:
+        - The name of the template in which the Port Channel Interface Policy Group has been created.
```
Comment (anvitha-jain) [inline plugins/modules/ndo_virtual_port_channel_interface.py:94]: Port Channel Interface Policy Group has been created -> Interface Policy Group has been  created.

**Review by anvitha-jain — CHANGES_REQUESTED** [review]

**Review by samiib — DISMISSED** [review]
LGTM

**Review by akinross — DISMISSED** [review]
LGTM

**Review by shrsr — APPROVED** [review]
LGTM

**Review by shrsr — DISMISSED** [review]
LGTM

**Review by sajagana — APPROVED** [review]
LGTM!

**Review by akinross — DISMISSED** [review]
LGTM

**Review by sajagana — APPROVED** [review]
LGTM!

**Review by shrsr — APPROVED** [review]
LGTM

**Review by samiib — APPROVED** [review]
LGTM

**Review by akinross — DISMISSED** [review]
LGTM

**Review by anvitha-jain — APPROVED** [review]
LGTM

**Comment by shrsr** [comment]
I'm curious as to what would happen in this scenario where node_1 is not mentioned in the second config and interfaces_1 is changed to 1/2? Likely node_1 would be sent in the payload as null and NDO would complain in the response? If yes, then we need to modify the append_update_ops_data() a little...
```
- name: Create a new virtual port channel interface 
      cisco.mso.ndo_virtual_port_channel_interface: 
        <<: *mso_info
        template: ansible_fabric_resource_template
        virtual_port_channel_interface: ansible_virtual_port_channel_interface
        description: Ansible Virtual Port Channel test
        node_1: 101
        node_2: 102
        interfaces_1: 1/1
        interfaces_2: 1/1
        interface_policy_group: 
          name: ansible_test_interface_policy_group_port_channel
          template: ansible_fabric_policy_template
        interface_descriptions:
          - node: 101
            interface_id: 1/1
            description: first Ansible interface test for first node
          - node: 102
            interface_id: 1/1
            description: first Ansible interface test for second node
        state: present

- name: Create a new virtual port channel interface with change in interfaces
      cisco.mso.ndo_virtual_port_channel_interface: 
        <<: *mso_info
        template: ansible_fabric_resource_template
        virtual_port_channel_interface: ansible_virtual_port_channel_interface
        description: Ansible Virtual Port Channel test
        node_2: 102
        interfaces_1: 1/2
        interfaces_2: 1/1
        interface_policy_group: 
          name: ansible_test_interface_policy_group_port_channel
          template: ansible_fabric_policy_template
        interface_descriptions:
          - node: 101
            interface_id: 1/1
            description: first Ansible interface test for first node
          - node: 102
            interface_id: 1/1
            description: first Ansible interface test for second node
        state: present
  ```

**Review by shrsr — CHANGES_REQUESTED** [review]
I'm curious as to what would happen in this scenario where node_1 is not mentioned in the second config and interfaces_1 is changed to 1/2? Likely node_1 would be sent in the payload as null and NDO would complain in the response? If yes, then we need to modify the append_update_ops_data() a little...
```
- name: Create a new virtual port channel interface 
      cisco.mso.ndo_virtual_port_channel_interface: 
        <<: *mso_info
        template: ansible_fabric_resource_template
        virtual_port_channel_interface: ansible_virtual_port_channel_interface
        description: Ansible Virtual Port Channel test
        node_1: 101
        node_2: 102
        interfaces_1: 1/1
        interfaces_2: 1/1
        interface_policy_group: 
          name: ansible_test_interface_policy_group_port_channel
          template: ansible_fabric_policy_template
        interface_descriptions:
          - node: 101
            interface_id: 1/1
            description: first Ansible interface test for first node
          - node: 102
            interface_id: 1/1
            description: first Ansible interface test for second node
        state: present

- name: Create a new virtual port channel interface with change in interfaces
      cisco.mso.ndo_virtual_port_channel_interface: 
        <<: *mso_info
        template: ansible_fabric_resource_template
        virtual_port_channel_interface: ansible_virtual_port_channel_interface
        description: Ansible Virtual Port Channel test
        node_2: 102
        interfaces_1: 1/2
        interfaces_2: 1/1
        interface_policy_group: 
          name: ansible_test_interface_policy_group_port_channel
          template: ansible_fabric_policy_template
        interface_descriptions:
          - node: 101
            interface_id: 1/1
            description: first Ansible interface test for first node
          - node: 102
            interface_id: 1/1
            description: first Ansible interface test for second node
        state: present
  ```

File: plugins/modules/ndo_virtual_port_channel_interface.py [hunk 14]
```diff
@@ -0,0 +1,443 @@
...
+            }
+            append_update_ops_data(ops, match.details, virtual_port_channel_attrs_path, mso_values)
+            mso.sanitize(match.details, collate=True)
+
+        else:
+            mso_values = dict(
+                name=name,
+                description=description,
+                node1Details=dict(
+                    node=node_1,
```
Comment (sajagana) [inline plugins/modules/ndo_virtual_port_channel_interface.py:407]: When you do not provide `node_1` and `node_2`. I got the below value, and the task failed. Also, check the update part.

```
{
    "op": "add",
    "path": "/fabricResourceTemplate/template/virtualPortChannels/-",
    "value": {
        "description": "Ansible Virtual Port Channel test",
        "interfaceDescriptions": [],
        "name": "ansible_virtual_port_channel_interface",
        "node1Details": {
            "memberInterfaces": "1/1",
            "node": null
        },
        "node2Details": {
            "memberInterfaces": "1/1",
            "node": null
        }
    }
}
```

File: tests/integration/targets/ndo_virtual_port_channel_interface/tasks/main.yml [hunk 15]
```diff
@@ -0,0 +1,563 @@
...
+    # CREATE
+
+    - name: Create a new virtual port channel interface (check_mode)
+      cisco.mso.ndo_virtual_port_channel_interface: &create_virtual_port_channel_interface
+        <<: *mso_info
+        template: ansible_fabric_resource_template
+        virtual_port_channel_interface: ansible_virtual_port_channel_interface
+        description: Ansible Virtual Port Channel test
+        node_1: 101
+        node_2: 102
```
Comment (sajagana) [inline tests/integration/targets/ndo_virtual_port_channel_interface/tasks/main.yml:93]: Add a create task without proving `node_1` and `node_2`.

**Review by sajagana — CHANGES_REQUESTED** [review]

File: plugins/modules/ndo_virtual_port_channel_interface.py (hunk 14 above)
Comment (gmicol) [reply plugins/modules/ndo_virtual_port_channel_interface.py:407]: By doing the same thing, I did not get any error or task failed, even try on the UI by creating a vpc without nodes 1 and 2 and still got no error. On which NDO version did you try @sajagana ? I tried on 4.4.1

File: tests/integration/targets/ndo_virtual_port_channel_interface/tasks/main.yml (hunk 15 above)
Comment (gmicol) [reply tests/integration/targets/ndo_virtual_port_channel_interface/tasks/main.yml:93]: Why should we add a test for this?

File: plugins/modules/ndo_virtual_port_channel_interface.py [hunk 16]
```diff
@@ -0,0 +1,443 @@
...
+        if uuid and not mso.existing:
+            mso.fail_json(msg="{0} with the UUID: '{1}' not found".format(object_description, uuid))
+
+        if interface_policy_group and not interface_policy_group_uuid:
+            fabric_policy_template = MSOTemplate(mso, "fabric_policy", interface_policy_group.get("template"))
+            fabric_policy_template.validate_template("fabricPolicy")
+            interface_policy_group_uuid = fabric_policy_template.get_interface_policy_group_uuid(interface_policy_group.get("name"))
+
+        if mso.existing:
+            mso_values = {
```
[the remark this answers is not in the document]
Comment (gmicol) [reply plugins/modules/ndo_virtual_port_channel_interface.py:389]: Should we in the case we have tuples as keys? I think it is better in all aspect to use `{}` format instead of using the dict() class for this case specifically because we would have to do something like this with the dict class:
```PYTHON
my_dict_test = dict([((1, 2), "value1"), ((3, 4), "value2")])
```
which is definitely less readable and less optimal than `{}` format.

File: tests/integration/targets/ndo_virtual_port_channel_interface/tasks/main.yml (hunk 15 above)
Comment (sajagana) [reply tests/integration/targets/ndo_virtual_port_channel_interface/tasks/main.yml:93]: The `node_1` and `node_2` are optional attributes. So we can add a task to cover that part.

File: plugins/modules/ndo_virtual_port_channel_interface.py (hunk 14 above)
Comment (sajagana) [reply plugins/modules/ndo_virtual_port_channel_interface.py:407]: Yes, it worked!

File: tests/integration/targets/ndo_virtual_port_channel_interface/tasks/main.yml (hunk 15 above)
Comment (anvitha-jain) [reply tests/integration/targets/ndo_virtual_port_channel_interface/tasks/main.yml:93]: @gmicol could you please add a test case which covers the above, since UI lets you do it without any issues.

**Review by anvitha-jain — CHANGES_REQUESTED** [review]

**Review by akinross — APPROVED** [review]
LGTM

**Review by shrsr — DISMISSED** [review]
LGTM

**Review by anvitha-jain — APPROVED** [review]
LGTM

**Review by akinross — DISMISSED** [review]
LGTM

**Review by sajagana — APPROVED** [review]
LGTM!

**Review by shrsr — DISMISSED** [review]
LGTM

**Review by anvitha-jain — DISMISSED** [review]
LGTM

**Review by akinross — DISMISSED** [review]
LGTM

**Review by samiib — CHANGES_REQUESTED** [review]

File: plugins/modules/ndo_virtual_port_channel_interface.py [hunk 17]
```diff
@@ -0,0 +1,443 @@
...
+    - This parameter is required when creating a new Virtual Port Channel Interface.
+    type: list
+    elements: str
+    aliases: [ interfaces_1, members_1 ]
+  interfaces_node_2:
+    description:
+    - The list of used Interface IDs for the second node.
+    - Ranges of Interface IDs can be used.
+    - This parameter is required when creating a new Virtual Port Channel Interface.
+    - If O(interfaces_node_2=[mirror]) is defined as an empty list and O(interfaces_node_1) is clearly defined,
```
[the remark this answers is not in the document]
Comment (gmicol) [reply plugins/modules/ndo_virtual_port_channel_interface.py:72]: you're right I forgot to remove that part, Thanks!

**Review by samiib — DISMISSED** [review]
LGTM

**Review by akinross — DISMISSED** [review]
LGTM

**Review by shrsr — DISMISSED** [review]
LGTM

**Review by anvitha-jain — DISMISSED** [review]

**Review by shrsr — APPROVED** [review]
LGTM

**Review by shrsr — APPROVED** [review]
LGTM

**Review by anvitha-jain — APPROVED** [review]
LGTM

**Review by samiib — APPROVED** [review]
LGTM

**Review by sajagana — APPROVED** [review]
LGTM!

**Review by akinross — APPROVED** [review]
LGTM

**Review by lhercot — APPROVED** [review]
LGTM
